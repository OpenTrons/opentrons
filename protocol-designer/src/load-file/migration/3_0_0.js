// @flow
import assert from 'assert'
import mapValues from 'lodash/mapValues'
import omit from 'lodash/omit'
import uniq from 'lodash/uniq'
import type {
  ProtocolFile,
  FileLabware,
  FilePipette,
} from '@opentrons/shared-data/protocol/flowTypes/schemaV3'
import { getAllDefinitions } from '../../labware-defs/utils'
import type { PDProtocolFile as PDProtocolFileV1, PDMetadata } from './1_1_0'

// NOTE: PDMetadata type did not change btw 1.1.0 and 3.0.0
export type PDProtocolFile = ProtocolFile<PDMetadata>

// the version and schema for this migration
export const PD_VERSION = '3.0.0'
export const SCHEMA_VERSION = 3

// TODO IMMEDIATELY review and confirm all mappings
const modelToURIMap = {
  '12-well-plate': 'corning_12_wellplate_6.9ml_flat',
  '24-well-plate': 'corning_24_wellplate_3.4ml_flat',
  '384-plate': 'corning_384_wellplate_112ul_flat',
  '48-well-plate': 'corning_48_wellplate_1.6ml_flat',
  '6-well-plate': 'corning_6_wellplate_16.8ml_flat',
  '96-PCR-flat': 'biorad_96_wellplate_200ul_pcr',
  '96-flat': 'generic_96_wellplate_340ul_flat',
  'biorad-hardshell-96-PCR': 'biorad_96_wellplate_200ul_pcr',
  'opentrons-aluminum-block-2ml-eppendorf':
    'opentrons_24_aluminumblock_generic_2ml_screwcap',
  'opentrons-aluminum-block-2ml-screwcap':
    'opentrons_24_aluminumblock_generic_2ml_screwcap',
  'opentrons-aluminum-block-96-PCR-plate':
    'opentrons_96_aluminumblock_biorad_wellplate_200ul',
  'opentrons-tiprack-300ul': 'opentrons_96_tiprack_300ul',
  'opentrons-tuberack-1.5ml-eppendorf':
    'opentrons_24_tuberack_eppendorf_1.5ml_safelock_snapcap',
  'opentrons-tuberack-15_50ml':
    'opentrons_10_tuberack_falcon_4x50ml_6x15ml_conical',
  'opentrons-tuberack-15ml': 'opentrons_15_tuberack_falcon_15ml_conical',
  'opentrons-tuberack-2ml-eppendorf':
    'opentrons_24_tuberack_eppendorf_2ml_safelock_snapcap',
  'opentrons-tuberack-2ml-screwcap':
    'opentrons_24_tuberack_generic_2ml_screwcap',
  'opentrons-tuberack-50ml': 'opentrons_6_tuberack_falcon_50ml_conical',
  'tiprack-1000ul': 'opentrons_96_tiprack_1000ul',
  'tiprack-10ul': 'opentrons_96_tiprack_10ul',
  'trough-12row': 'usascientific_12_reservoir_22ml',
  'fixed-trash': 'opentrons_1_trash_1100ml_fixed',
}

function v1LabwareModelToV2URI(model: string): string {
  const uri = `opentrons/${modelToURIMap[model]}/1`
  assert(uri, `expected a v2 URI for v1 labware model "${model}"`)
  return uri
}

function migrateMetadata(
  metadata: $PropertyType<PDProtocolFileV1, 'metadata'>
): $PropertyType<PDProtocolFile, 'metadata'> {
  const metadataExtraKeys: typeof metadata = {
    ...metadata,
    protocolName: metadata['protocol-name'],
    lastModified: metadata['last-modified'],
  }
  return omit(metadataExtraKeys, ['protocol-name', 'last-modified'])
}

type FilePipettes = $PropertyType<PDProtocolFile, 'pipettes'>
function migratePipettes(
  pipettes: $PropertyType<PDProtocolFileV1, 'pipettes'>
): FilePipettes {
  return Object.keys(pipettes).reduce<FilePipettes>(
    (acc, pipetteId: string) => {
      const oldPipette = pipettes[pipetteId]
      const nextPipette: FilePipette = {
        mount: oldPipette.mount,
        name: oldPipette.name || (oldPipette.model || '').split('_v')[0],
      }
      // NOTE: the hacky model to name ('p10_single_v1.3' -> 'p10_single') fallback
      // should already be handled in most cases, this is just for safety
      return { ...acc, [pipetteId]: nextPipette }
    },
    {}
  )
}

type FileLabwares = $PropertyType<PDProtocolFile, 'labware'>
function migrateLabware(
  labware: $PropertyType<PDProtocolFileV1, 'labware'>
): FileLabwares {
  return Object.keys(labware).reduce<FileLabwares>((acc, labwareId: string) => {
    const oldLabware = labware[labwareId]
    const nextLabware: FileLabware = {
      slot: oldLabware.slot,
      displayName: oldLabware['display-name'],
      definitionId: v1LabwareModelToV2URI(oldLabware.model),
    }
    return { ...acc, [labwareId]: nextLabware }
  }, {})
}

type LabwareDefinitions = $PropertyType<PDProtocolFile, 'labwareDefinitions'>
function getLabwareDefinitions(
  labware: $PropertyType<PDProtocolFileV1, 'labware'>
): LabwareDefinitions {
  const allLabwareModels = uniq(
    Object.keys(labware).map((labwareId: string) => labware[labwareId].model)
  )
  const allLabwareURIs = allLabwareModels.map(v1LabwareModelToV2URI)
  return allLabwareURIs.reduce<LabwareDefinitions>((acc, uri: string) => {
    const labwareDef = getAllDefinitions()[uri]
    assert(labwareDef, `could not find labware def for ${uri}`)
    return { ...acc, [uri]: labwareDef }
  }, {})
}

function migrateAppData(appData: PDMetadata): PDMetadata {
  // update pipette tiprack v1 model to v2 URI
  return {
    ...appData,
    pipetteTiprackAssignments: mapValues(
      appData.pipetteTiprackAssignments,
      model => v1LabwareModelToV2URI(model)
    ),
  }
}

const migrateFile = (fileData: PDProtocolFileV1): PDProtocolFile => {
  return {
    designerApplication: {
      name: 'opentrons/protocol-designer',
      version: PD_VERSION,
      data: migrateAppData(fileData['designer-application'].data),
    },
    schemaVersion: SCHEMA_VERSION,
    metadata: migrateMetadata(fileData.metadata),
    robot: fileData.robot,
    pipettes: migratePipettes(fileData.pipettes),
    labware: migrateLabware(fileData.labware),
    labwareDefinitions: getLabwareDefinitions(fileData.labware),
    commands: [], // NOTE: these will be generated by PD upon import
  }
}

export default migrateFile
