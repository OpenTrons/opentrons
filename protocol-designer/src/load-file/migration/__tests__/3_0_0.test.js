import { migrateFile } from '../3_0_0'
import example_1_1_0 from '../../../../fixtures/protocol/1/example_1_1_0.json'

jest.mock('../../../labware-defs/utils')
jest.mock('../utils/v1LabwareModelToV2Def')

describe('migrate to 3.0.0', () => {
  it('snapshot test', () => {
    const result = migrateFile(example_1_1_0)
    expect(result).toMatchSnapshot()
  })
})
