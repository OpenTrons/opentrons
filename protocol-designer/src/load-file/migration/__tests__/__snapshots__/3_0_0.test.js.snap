// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrate to 3.0.0 snapshot test 1`] = `
Object {
  "commands": Array [],
  "designerApplication": Object {
    "data": Object {
      "dismissedWarnings": Object {
        "form": Object {},
        "timeline": Object {},
      },
      "ingredLocations": Object {
        "dafd4000-92a5-11e9-ac62-1b173f839d9e:96-deep-well": Object {
          "A1": Object {
            "0": Object {
              "volume": 121,
            },
          },
          "B1": Object {
            "0": Object {
              "volume": 121,
            },
          },
          "C1": Object {
            "0": Object {
              "volume": 121,
            },
          },
          "D1": Object {
            "0": Object {
              "volume": 121,
            },
          },
          "E1": Object {
            "0": Object {
              "volume": 121,
            },
          },
          "F1": Object {
            "1": Object {
              "volume": 44,
            },
          },
          "G1": Object {
            "1": Object {
              "volume": 44,
            },
          },
          "H1": Object {
            "1": Object {
              "volume": 44,
            },
          },
        },
      },
      "ingredients": Object {
        "0": Object {
          "description": null,
          "liquidGroupId": "0",
          "name": "samples",
          "serialize": false,
        },
        "1": Object {
          "description": null,
          "liquidGroupId": "1",
          "name": "dna",
          "serialize": false,
        },
      },
      "orderedStepIds": Array [
        "e7d36200-92a5-11e9-ac62-1b173f839d9e",
        "18113c80-92a6-11e9-ac62-1b173f839d9e",
        "2e622080-92a6-11e9-ac62-1b173f839d9e",
      ],
      "pipetteTiprackAssignments": Object {
        "c6f45030-92a5-11e9-ac62-1b173f839d9e": "fixture/fixture_regular_example_1/1",
        "c6f47740-92a5-11e9-ac62-1b173f839d9e": "fixture/fixture_regular_example_1/1",
      },
      "savedStepForms": Object {
        "18113c80-92a6-11e9-ac62-1b173f839d9e": Object {
          "aspirate_flowRate": 8,
          "blowout_checkbox": true,
          "blowout_location": "dest_well",
          "changeTip": "always",
          "dispense_flowRate": 7,
          "id": "18113c80-92a6-11e9-ac62-1b173f839d9e",
          "labware": "dafd4000-92a5-11e9-ac62-1b173f839d9e:96-deep-well",
          "mix_mmFromBottom": 0.5,
          "mix_touchTip_checkbox": true,
          "mix_touchTip_mmFromBottom": 30.5,
          "mix_wellOrder_first": "t2b",
          "mix_wellOrder_second": "l2r",
          "pipette": "c6f45030-92a5-11e9-ac62-1b173f839d9e",
          "stepDetails": "",
          "stepName": "mix",
          "stepType": "mix",
          "times": 3,
          "volume": "5.5",
          "wells": Array [
            "F1",
          ],
        },
        "2e622080-92a6-11e9-ac62-1b173f839d9e": Object {
          "id": "2e622080-92a6-11e9-ac62-1b173f839d9e",
          "pauseForAmountOfTime": "true",
          "pauseHour": 1,
          "pauseMessage": "Delay plz",
          "pauseMinute": 2,
          "pauseSecond": 3,
          "stepDetails": "",
          "stepName": "pause",
          "stepType": "pause",
        },
        "__INITIAL_DECK_SETUP_STEP__": Object {
          "id": "__INITIAL_DECK_SETUP_STEP__",
          "labwareLocationUpdate": Object {
            "c6f4ec70-92a5-11e9-ac62-1b173f839d9e:tiprack-10ul": "1",
            "c6f51380-92a5-11e9-ac62-1b173f839d9e:tiprack-200ul": "2",
            "dafd4000-92a5-11e9-ac62-1b173f839d9e:96-deep-well": "10",
            "trashId": "12",
          },
          "pipetteLocationUpdate": Object {
            "c6f45030-92a5-11e9-ac62-1b173f839d9e": "left",
            "c6f47740-92a5-11e9-ac62-1b173f839d9e": "right",
          },
          "stepType": "manualIntervention",
        },
        "e7d36200-92a5-11e9-ac62-1b173f839d9e": Object {
          "aspirate_flowRate": 0.6,
          "aspirate_labware": "dafd4000-92a5-11e9-ac62-1b173f839d9e:96-deep-well",
          "aspirate_mix_checkbox": true,
          "aspirate_mix_times": 3,
          "aspirate_mix_volume": "2",
          "aspirate_mmFromBottom": 1,
          "aspirate_touchTip_checkbox": true,
          "aspirate_touchTip_mmFromBottom": 28.5,
          "aspirate_wellOrder_first": "t2b",
          "aspirate_wellOrder_second": "l2r",
          "aspirate_wells": Array [
            "A1",
          ],
          "aspirate_wells_grouped": false,
          "blowout_checkbox": true,
          "blowout_location": "trashId",
          "changeTip": "always",
          "dispense_labware": "dafd4000-92a5-11e9-ac62-1b173f839d9e:96-deep-well",
          "dispense_mix_checkbox": true,
          "dispense_mix_times": 2,
          "dispense_mix_volume": "3",
          "dispense_mmFromBottom": 2.5,
          "dispense_touchTip_checkbox": true,
          "dispense_wellOrder_first": "b2t",
          "dispense_wellOrder_second": "r2l",
          "dispense_wells": Array [
            "C6",
            "D6",
            "E6",
            "C7",
            "D7",
            "E7",
            "C8",
            "D8",
            "E8",
          ],
          "disposalVolume_checkbox": true,
          "disposalVolume_volume": "1",
          "id": "e7d36200-92a5-11e9-ac62-1b173f839d9e",
          "path": "single",
          "pipette": "c6f45030-92a5-11e9-ac62-1b173f839d9e",
          "preWetTip": false,
          "stepDetails": "yeah notes",
          "stepName": "transfer things",
          "stepType": "moveLiquid",
          "volume": "6",
        },
      },
    },
    "name": "opentrons/protocol-designer",
    "version": "3.0.0",
  },
  "labware": Object {
    "c6f4ec70-92a5-11e9-ac62-1b173f839d9e:tiprack-10ul": Object {
      "definitionId": "fixture/fixture_regular_example_1/1",
      "displayName": "tiprack 10ul (1)",
      "slot": "1",
    },
    "c6f51380-92a5-11e9-ac62-1b173f839d9e:tiprack-200ul": Object {
      "definitionId": "fixture/fixture_regular_example_1/1",
      "displayName": "tiprack 200ul (1)",
      "slot": "2",
    },
    "dafd4000-92a5-11e9-ac62-1b173f839d9e:96-deep-well": Object {
      "definitionId": "fixture/fixture_regular_example_1/1",
      "displayName": "96 deep well (1)",
      "slot": "10",
    },
    "trashId": Object {
      "definitionId": "fixture/fixture_regular_example_1/1",
      "displayName": "Trash",
      "slot": "12",
    },
  },
  "labwareDefinitions": Object {
    "fixture/fixture_regular_example_1/1": Object {
      "brand": Object {
        "brand": "opentrons",
        "brandId": Array [
          "t40u9sernisofsea",
        ],
      },
      "cornerOffsetFromSlot": Object {
        "x": 0,
        "y": 0,
        "z": 0,
      },
      "dimensions": Object {
        "xDimension": 127.76,
        "yDimension": 85.48,
        "zDimension": 50,
      },
      "groups": Array [
        Object {
          "metadata": Object {},
          "wells": Array [
            "A1",
            "A2",
          ],
        },
      ],
      "metadata": Object {
        "displayCategory": "wellPlate",
        "displayName": "fake labware",
        "displayVolumeUnits": "ÂµL",
      },
      "namespace": "fixture",
      "ordering": Array [
        Array [
          "A1",
        ],
        Array [
          "A2",
        ],
      ],
      "parameters": Object {
        "format": "96Standard",
        "isMagneticModuleCompatible": false,
        "isTiprack": false,
        "loadName": "fixture_regular_example_1",
      },
      "schemaVersion": 2,
      "version": 1,
      "wells": Object {
        "A1": Object {
          "depth": 40,
          "diameter": 30,
          "shape": "circular",
          "totalLiquidVolume": 100,
          "x": 10,
          "y": 75.48,
          "z": 15,
        },
        "A2": Object {
          "depth": 40,
          "diameter": 30,
          "shape": "circular",
          "totalLiquidVolume": 100,
          "x": 20,
          "y": 75.48,
          "z": 15,
        },
      },
    },
  },
  "metadata": Object {
    "author": "Author name",
    "category": null,
    "created": 1560957631666,
    "description": "Description here",
    "lastModified": undefined,
    "protocolName": "Some name!",
    "subcategory": null,
    "tags": Array [],
  },
  "pipettes": Object {
    "c6f45030-92a5-11e9-ac62-1b173f839d9e": Object {
      "mount": "left",
      "name": "p10_single",
    },
    "c6f47740-92a5-11e9-ac62-1b173f839d9e": Object {
      "mount": "right",
      "name": "p50_single",
    },
  },
  "robot": Object {
    "model": "OT-2 Standard",
  },
  "schemaVersion": 3,
}
`;
